src/testnet/tenure.rs

```diff
@@ -0,0 +1,105 @@
+    started_at: std::time::Instant,
+    vrf_seed: VRFSeed,
+}
+
+impl <'a> LeaderTenure {
+
+    pub fn new(parent_block: StacksHeaderInfo, 
+               average_block_time: u64,
+               coinbase_tx: StacksTransaction,
+               config: NodeConfig,
+               mem_pool: MemPoolFS,
+               microblock_secret_key: StacksPrivateKey,  
+               last_sortitioned_block: SortitionedBlock,
+               vrf_proof: VRFProof) -> LeaderTenure {
+
+        let now = time::Instant::now();
+
+        let ratio = StacksWorkScore {
+            burn: parent_block.anchored_header.total_work.burn + 1,
+            work: parent_block.anchored_header.total_work.work + 1,
+        };
+
+        let block_builder = match last_sortitioned_block.block_height {
+            0 => StacksBlockBuilder::first(1, &parent_block.burn_header_hash, &vrf_proof, &microblock_secret_key),
+            _ => StacksBlockBuilder::from_parent(1, &parent_block, &ratio, &vrf_proof, &microblock_secret_key)
+        };
+
+        Self {
+            average_block_time,
+            block_builder,
+            coinbase_tx,
+            config,
+            last_sortitioned_block,
+            mem_pool,
+            parent_block,
+            started_at: now,
+            vrf_seed: VRFSeed::from_proof(&vrf_proof),
+        }
+    }
+
+    pub fn handle_txs(&mut self, clarity_tx: &mut ClarityTx<'a>, txs: Vec<StacksTransaction>) {
+        for tx in txs {
+            self.block_builder.try_mine_tx(clarity_tx, &tx).unwrap();
+        }
+    }
+
+    pub fn run(&mut self) -> Option<(StacksBlock, Vec<StacksMicroblock>, SortitionedBlock)> {
+
+        let mut chain_state = StacksChainState::open(false, 0x80000000, &self.config.path).unwrap();
+
+        let mut clarity_tx = match self.last_sortitioned_block.block_height {
+            0 => {
+                chain_state.block_begin(
+                &BurnchainHeaderHash([0u8; 32]),
+                &BlockHeaderHash([0u8; 32]),
+                &BurnchainHeaderHash([1u8; 32]), 
+                &BlockHeaderHash([1u8; 32]))
+            },
+            _ => chain_state.block_begin(
+                &self.last_sortitioned_block.burn_header_hash, 
+                &self.parent_block.anchored_header.block_hash(), 
+                &BurnchainHeaderHash([1u8; 32]), 
+                &BlockHeaderHash([1u8; 32])),
+        };
+
+        let mempool_poll_interval = time::Duration::from_millis(250);
+        let tenure_duration = time::Duration::from_millis(self.average_block_time * 1 / 2);
+        let should_commit_block_at = self.started_at.checked_add(tenure_duration).unwrap();
+
+        self.handle_txs(&mut clarity_tx, vec![self.coinbase_tx.clone()]);
+
+        while time::Instant::now() < should_commit_block_at {
+            let txs = self.mem_pool.poll();
+            self.handle_txs(&mut clarity_tx, txs);
+            thread::sleep(mempool_poll_interval);
+        }
+
+        let b = self.block_builder.clone();
```

Comment:
Why are we `clone()`ing this?  You should reset the `block_builder` on each pass of this loop -- it's meant to be instantiated only once per epoch.

Reply:
> Why are we clone()ing this?
This was a leftover. Addressed with https://github.com/blockstack/blockstack-core/pull/1202/commits/fc2a33d24f3cbea59b96ed7580741122e32d4dde

> You should reset the block_builder
I think this is the case. Once the tenure ran and has been processed, it's getting released.